src/Markdig/Helpers/CustomArrayPool.cs

```diff
@@ -0,0 +1,86 @@
+using System.Threading;
+
+namespace Markdig.Helpers
+{
+    internal sealed class CustomArrayPool<T>
+    {
+        private sealed class Bucket
+        {
+            private readonly T[][] _buffers;
+
+            private int _index;
+            private int _lock;
+
+            public Bucket(int numberOfBuffers)
+            {
+                _buffers = new T[numberOfBuffers][];
+            }
+
+            public T[] Rent()
+            {
+                T[][] buffers = _buffers;
+                T[] buffer = null;
+                if (Interlocked.CompareExchange(ref _lock, 1, 0) == 0)
```

Comment:
Are you sure that this code is thread safe?
If we have 2 threads entering this method at the exact same time, thread-1 enter this line (so _lock is then set to 1), so thread-2 can't enter (_lock is 1), so it goes directly to `return buffer` which will be null, no?

Reply:
(this case is probably a better fit for a slim mutex)

Reply:
This is heavily adapted from the [CoreFx's ConfigurableArrayPool](https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Buffers/ConfigurableArrayPool.cs#L163). A SpinLock is used there instead.

I expect very little contention on this lock. Parsing is mostly done by one thread and even then dealing with the StringLine buffers all that common.
As you said, in this implementation if there is contention, the second thread will return null. That will fallback to a regular `new T[length]` allocation. Such a buffer can still be returned to the pool later on.

The reason I opted for that is because not all buffers will get released. In the case of StringLineGroup, only objects where inlines are processed will release the buffer in the end. In other cases the Lines are still available on the AST and adding a finalizer to reclaim them is way to expensive.

As such only buffers from objects that have their ProcessInlines set will get reclaimed. For other objects, buffers will only be returned when resizing, the final buffer on them will be reclaimed by the GC eventually and not returned to the pool.

Reply:
oh, ok, sorry, didn't look further with the `return SelectBucket(length)?.Rent() ?? new T[length];`
yeah that's ok then

Reply:
In other words, buffers aren't guaranteed to be returned from the actual bucket, they can also be allocated on the spot if the bucket is contended/at capacity/of incorrect size.
Over-allocating here is fine since some buffers will never get returned anyway.